prov/rxm/src/rxm_conn.c

```diff
@@ -1361,18 +1403,123 @@ static void *rxm_conn_progress(void *arg)
+}
+
+/* The parameter entry is cleared upon success, it should be cleared prior
+ * to this first call to this function. */
+static inline int
+rxm_conn_auto_progress_eq(struct rxm_ep *rxm_ep, struct rxm_msg_eq_entry *entry)
+{
+	while (1) {
+		entry->rd = rxm_eq_read(rxm_ep, RXM_CM_ENTRY_SZ, entry);
+		if (OFI_UNLIKELY(!entry->rd || entry->rd == -FI_EAGAIN))
+			return FI_SUCCESS;
+		if (entry->rd < 0 &&
+		    entry->rd != -FI_ECONNREFUSED)
+			break;
+		if (rxm_conn_eq_event(rxm_ep, entry))
+			break;
+		memset(entry, 0, RXM_MSG_EQ_ENTRY_SZ);
+	}
+	return -1;
+}
+
+static void *rxm_conn_atomic_progress(void *arg)
+{
+	struct rxm_ep *rxm_ep = container_of(arg, struct rxm_ep, util_ep);
+	struct rxm_fabric *rxm_fabric;
+	struct fid *fids[2];
+	struct pollfd fds[2];
+	struct rxm_msg_eq_entry *entry;
+	int again;
+	int ret;
+
+	assert(rxm_ep->msg_eq);
+	rxm_fabric = container_of(rxm_ep->util_ep.domain->fabric,
+				  struct rxm_fabric, util_fabric);
+	entry = calloc(1, RXM_MSG_EQ_ENTRY_SZ);
+	if (!entry) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to allocate memory!\n");
+		return NULL;
+	}
+
+	fids[0] = &rxm_ep->msg_eq->fid;
+	ret = fi_control(fids[0], FI_GETWAIT, (void *) &fds[0].fd);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to get MSG_EP EQ WAIT_FD %d\n", ret);
+		goto exit;
+	}
+	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
+	       "Starting CM conn thread with atomic AUTO_PROGRESS\n");
+	fds[0].events = POLLIN;
+
+	ofi_ep_lock_acquire(&rxm_ep->util_ep);
+	while (1) {
+		/* CQ can be bound after progress thread is started */
+		if (rxm_ep->msg_cq)
+			break;
+		again = fi_trywait(rxm_fabric->msg_fabric, fids, 1);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
+
+		if (!again) {
+			fds[0].revents = 0;
+
+			ret = poll(fds, 1, -1);
+			if (OFI_UNLIKELY(ret == -1)) {
+				if (errno == EINTR)
+					continue;
+				FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+					"Select error %d, closing CM thread\n",
+					errno);
+				goto exit;
+			}
 		}
-		ret = rxm_conn_handle_event(rxm_ep, entry);
-		if (ret)
+		if ( rxm_conn_auto_progress_eq(rxm_ep, entry))
 			goto exit;
-		memset(entry, 0, RXM_MSG_EQ_ENTRY_SZ);
+
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
+	}
+
+	/* CQ has been created, include it in poll */
+	fids[1] = &rxm_ep->msg_cq->fid;
+	fds[1].fd = rxm_ep->msg_cq_fd;
+	fds[1].events = POLLIN;
+
+	while(1) {
+		again = fi_trywait(rxm_fabric->msg_fabric, fids, 2);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
+
+		if (!again) {
+			fds[0].revents = 0;
+			fds[1].revents = 0;
+
+			ret = poll(fds, 2, -1);
+			if (OFI_UNLIKELY(ret == -1)) {
+				if (errno == EINTR)
+					continue;
+				FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+					"Select error %d, closing CM thread\n",
+					errno);
+				goto exit;
+			}
+		}
+		if (again || fds[0].revents & POLLIN) {
+			if ( rxm_conn_auto_progress_eq(rxm_ep, entry))
+				goto exit;
+		}
+		if (again || fds[1].revents & POLLIN)
+			rxm_ep_progress(&rxm_ep->util_ep);
+
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
```

Comment:
This function is a little long and has 2 while(1) loops.  Something in this structure needs updating, especially to avoid 2 infinite loops.

Reply:
It was originally only 1 loop; @a-ilango thought (and I agreed) that it was better to break it into two different loops so that once the CQ has been created and bound we don't have to do checks to see if it exists which each loop. With that said, I can break it into two functions if you still prefer.

Reply:
@shefty, @a-ilango - I broke the atomic auto progress EQ only piece out into a separate function. Once the CQ is bound, that function will terminate, and then the loop for the EQ and CQ will be initiated. Thanks for your comments/suggestions.